File: src/py_action/py_action/move_robot_server.py
#!/usr/bin/env python3
import rclpy
import time
import threading
from rclpy.node import Node
from rclpy.action import ActionServer
from rclpy.action.server import ServerGoalHandle, GoalResponse, CancelResponse
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup
from my_robot_interfaces.action import MoveRobot
 
 
class MoveRobotServerNode(Node): 
    def __init__(self):
        super().__init__("move_robot_server") 
        self.robot_position_=50
        self.goal_handle_:ServerGoalHandle = None
        self.goal_lock_ = threading.Lock()
        self.move_robot_server_= ActionServer(
            self,
            MoveRobot,
            "move_robot",
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback,
            execute_callback=self.execute_callback,
            callback_group=ReentrantCallbackGroup()
        )
        self.get_logger().info("Action Server has been started")
        self.get_logger().info("Robot position: "+ str(self.robot_position_))

    def cancel_callback(self, goal_handle:CancelResponse):
        self.get_logger().info("Received the cancel request")
        return CancelResponse.ACCEPT

    def goal_callback(self, goal_request: MoveRobot.Goal):
        self.get_logger().info("Received a new goal")
        if(goal_request._position not in range (0,100)) or goal_request.velocity <=0:
            self.get_logger().error("Rejecting the goal: Invalid position or velocity")
            return GoalResponse.REJECT
        
        #New goal is valid, abort previous goal and accept new goal
        if self.goal_handle_ is not None and self.goal_handle_.is_active:
            self.goal_handle_.abort()

        self.get_logger().info("Accepted the goal")
        return GoalResponse.ACCEPT

    def execute_callback(self, goal_handle: ServerGoalHandle):
        with self.goal_lock_:
            self.goal_handle_ = goal_handle

        goal_position = goal_handle.request.position
        goal_velocity = goal_handle.request.velocity

        self.get_logger().info("Executing the goal")

        result = MoveRobot.Result()
        feedback = MoveRobot.Feedback()

        while rclpy.ok():
            #Return if goal is rejected
            if not goal_handle.is_active:
                result.position = self.robot_position_
                result.message = "Pre-empted by another goal"
                return result
            
            #Cancel if requested
            if goal_handle.is_cancel_requested:
                result.position = self.robot_position_
                if goal_position == self.robot_position_:
                    result.message = "Succeed"
                    goal_handle.succeed()
                    return result
                result.message = "Cancelled by client"
                goal_handle.canceled()
                return result

            diff = goal_position - self.robot_position_
            if diff == 0:
                result.position = self.robot_position_
                result.message = "Succeed"
                goal_handle.succeed()
                return result
            elif diff > 0:
                if diff>=goal_velocity:
                    self.robot_position_+=goal_velocity
                else:
                    self.robot_position_+=diff
            else:
                if abs(diff)>=goal_velocity:
                    self.robot_position_-=goal_velocity
                else:
                    self.robot_position_+=diff
            feedback.current_position = self.robot_position_
            goal_handle.publish_feedback(feedback)
            self.get_logger().info("Robot position: "+ str(self.robot_position_))
            time.sleep(1.0)
 
 
def main(args=None):
    rclpy.init(args=args)
    node = MoveRobotServerNode() 
    rclpy.spin(node, MultiThreadedExecutor())
    rclpy.shutdown()
 
 
if __name__ == "__main__":
    main()

File: ros2_ws/src/cpp_action/src/move_robot_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"


#include "my_robot_interfaces/action/move_robot.hpp"

using MoveRobot = my_robot_interfaces::action::MoveRobot;
using MoveRobotGoalHandle = rclcpp_action::ServerGoalHandle<MoveRobot>;
using namespace std::placeholders;
 
class MoveRobotServerNode : public rclcpp::Node 
{
public:
    MoveRobotServerNode() : Node("move_robot_server") 
    {
        robot_position_ = 50;
        call_back_group_ = create_callback_group(rclcpp::CallbackGroupType::Reentrant);
        move_robot_server_ = rclcpp_action::create_server<MoveRobot>(
            this,
            "move_robot",
            std::bind(&MoveRobotServerNode::goal_callback, this, _1, _2),
            std::bind(&MoveRobotServerNode::cancel_callback, this, _1),
            std::bind(&MoveRobotServerNode::handle_accepted_callback_callback, this, _1),
            rcl_action_server_get_default_options(),
            call_back_group_
        );
        RCLCPP_INFO(get_logger(), "Action Server has been started");
    }
 
private:
    int robot_position_;
    std::mutex mutex_;
    std::shared_ptr<MoveRobotGoalHandle> goal_handle_;
    
    rclcpp::CallbackGroup::SharedPtr call_back_group_;
    rclcpp_action::Server<MoveRobot>::SharedPtr move_robot_server_;
    rclcpp_action::GoalUUID preempted_goal_id;

    rclcpp_action::GoalResponse goal_callback(
        const rclcpp_action::GoalUUID &uuid,
        std::shared_ptr<const MoveRobot::Goal> goal){
            RCLCPP_INFO(get_logger(), "Received a goal");

            //Validate new goal
            if(goal->position < 0 or goal->position >100 ){
                RCLCPP_WARN(get_logger(), "Position is not in the range (0,100) \nRejecting the goal");
                return rclcpp_action::GoalResponse::REJECT;
            }
            if(goal->velocity <= 0 ){
                RCLCPP_WARN(get_logger(), "Velcoity is less than 0 \nRejecting the goal");
                return rclcpp_action::GoalResponse::REJECT;
            }
            
            //Pre-empt any existing goals
            {
                std::lock_guard<std::mutex> lock(mutex_);
                if(goal_handle_){
                    if(goal_handle_->is_active()){
                        preempted_goal_id = goal_handle_->get_goal_id();
                    }
                }
            }

            RCLCPP_INFO(get_logger(), "Accepted the goal");
            return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
        }

    rclcpp_action::CancelResponse cancel_callback(
        const std::shared_ptr<MoveRobotGoalHandle> goal_handle){
            (void) goal_handle;
            RCLCPP_INFO(get_logger(), "Received the cancelRequest");
            return rclcpp_action::CancelResponse::ACCEPT;
        }

    void handle_accepted_callback_callback(
        const std::shared_ptr<MoveRobotGoalHandle> goal_handle){
            RCLCPP_INFO(get_logger(), "Executing the goal");
            execute_goal(goal_handle);
        }

    void execute_goal(
        const std::shared_ptr<MoveRobotGoalHandle> goal_handle){
            
            {
                std::lock_guard<std::mutex> lock(mutex_);
                goal_handle_ = goal_handle;
            }

            //Get request from the goal
            int target_position = goal_handle->get_goal()->position;
            int velocity = goal_handle->get_goal()->velocity;

            RCLCPP_INFO(get_logger(), "Target robot position: %d", int(target_position));
            RCLCPP_INFO(get_logger(), "Velocity : %d", int(velocity));
            RCLCPP_INFO(get_logger(), "Current robot position: %d", int(robot_position_));

            auto result = std::make_shared<MoveRobot::Result>();
            auto feedback = std::make_shared<MoveRobot::Feedback>();

            rclcpp::Rate loop_rate(1.0);
            while(rclcpp::ok()){
                //Check if pre-empt required
                {
                    std::lock_guard<std::mutex> lock(mutex_);
                        if(goal_handle->get_goal_id() == preempted_goal_id){
                            result->position = robot_position_;
                            result->message = "Pre-empted by another goal";
                            goal_handle->abort(result);
                            return;
                        }
                }

                //Check if cancelling
                if(goal_handle->is_canceling()){
                    result->position = robot_position_;
                    if(target_position==robot_position_){
                        result->message = "Succeed";
                        goal_handle->succeed(result);
                    }
                    else{
                        result->message = "Cancelled";
                        goal_handle->canceled(result);
                    }
                    return;
                }
                

                auto diff = target_position - robot_position_;

                if (diff == 0){
                    result->position = robot_position_;
                    result->message = "Succeed";
                    goal_handle->succeed(result);
                    return;
                }
                
                else if (diff > 0){
                    if (diff>=velocity)
                        robot_position_+=velocity;
                    else
                        robot_position_+=diff;
                }
                
                else{
                    if (abs(diff)>=velocity)
                        robot_position_-=velocity;
                    else
                        robot_position_+=diff;
                }

                loop_rate.sleep();
                feedback->current_position = robot_position_;
                goal_handle->publish_feedback(feedback);
                RCLCPP_INFO(get_logger(), "Current robot position: %d", int(robot_position_));
            }
    }
};
 
int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MoveRobotServerNode>(); 
    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node);
    executor.spin();
    rclcpp::shutdown();
    return 0;
}